lib/IDE/Refactoring.cpp

```diff
@@ -1453,6 +1456,105 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
                                           EditConsumer).performChange();
 }
 
+/// Compute a context C such that C is a parent context of A and B.
+/// If no such context exists, return \c nullptr.
+DeclContext *getCommonParentContext(DeclContext *A, DeclContext *B) {
+  if (A == B)
+    return A;
+
+  if (A->isChildContextOf(B))
+    return B;
+
+  if (B->isChildContextOf(A))
+    return A;
+
+  // Peel away layers of A until we reach a common parent
+  for (DeclContext *CurDC = A; CurDC; CurDC = CurDC->getParent()) {
+    if (B->isChildContextOf(CurDC))
+      return CurDC;
+  }
+
+  return nullptr;
+}
+
+bool RefactoringActionMoveMembersToExtension::isApplicable(
+    ResolvedRangeInfo Info, DiagnosticEngine &Diag) {
+  switch (Info.Kind) {
+  case RangeKind::SingleDecl:
+  case RangeKind::MultiDecl: {
+    // Compute a decl context that is the parent context for all selected
+    // decls
+    DeclContext *CommonDC = nullptr;
+    for (auto DD : Info.DeclaredDecls) {
+      if (!CommonDC) {
+        CommonDC = DD.VD->getDeclContext();
+      } else {
+        CommonDC = getCommonParentContext(CommonDC, DD.VD->getDeclContext());
+      }
+    }
```

Comment:
Could we move this for loop to its own function to eliminate the code duplication with the `performChange` part?